### PR TITLE
Add validation to transition type

### PR DESCRIPTION
**Pull request by elyscape**
This adds validations to ensure that:

1. The value of `resource` is either a valid resource reference or a string that can be converted to a valid resource reference.
2. The resource specified in `resource` exists in the catalog.
3. The value of `attributes` is a hash.
4. All the specified attributes are valid parameters for the specified resource.
5. The values of `prior_to` are valid resource references or strings that can be converted to valid resource references.
6. All resources specified in `prior_to` exist in the catalog.

Additionally, it adds munging to resolve the values of `resource` and `prior_to` into their catalog representations and removes the now-redundant resource resolution from the provider.

### CODE REVIEW

**Comment by elyscape** [comment]
Any update on this?

**Comment by elyscape** [comment]
Pinging @hunner, @reidmv.

**Comment by reidmv** [comment]
@elyscape whoa, thanks for doing this! This didn't get on my radar. Sorry for the radio silence on the PR.

I really like the general form of this. It looks like there are two problems that need to be resolved before it can be merged. I'll comment inline on where I'm having problems running the new code.

File: lib/puppet/type/transition.rb [hunk 1]
```diff
@@ -86,8 +81,15 @@ def insync?(is)
...
 
     validate do |value|
-      # TODO: validate that this is a hash.
-      true
+      unless value.is_a? Hash
+        raise ArgumentError, "#{value} is not a hash"
+      end
+      res = @resource[:resource]
+      value.keys.each do |attribute|
+        unless res.valid_parameter?(attribute)
```
Comment (reidmv) [inline lib/puppet/type/transition.rb]: Running under Puppet 4.3.2.

Manifest:

```puppet
file { '/tmp/test':
  ensure  => file,
  content => 'enabled=1',
  before  => Notify['test'],
}

notify { 'test': }

transition { 'stop myapp service':
  resource   => File['/tmp/test'],
  attributes => { ensure => absent },
  prior_to   => Notify['test']
}
```

Error:
```
[reidmv@halcyon:~/src/puppetlabs-transition/] % puppet apply test2.pp
Notice: Compiled catalog for halcyon.corp.puppetlabs.net in environment production in 0.05 seconds
Error: Parameter attributes failed on Transition[stop myapp service]: Validate method failed for class attributes: undefined method `valid_parameter?' for #<Puppet::Type::File:0x007faca4851788> at /Users/reidmv/src/puppetlabs-transition/test2.pp:9
```

File: lib/puppet/type/transition.rb [hunk 2]
```diff
@@ -136,4 +147,29 @@ def autorequire(rel_catalog = nil)
...
+      rescue ArgumentError
+        raise ArgumentError, "#{res} is not a valid resource reference"
+      end
+    else
+      raise ArgumentError, "#{res} is not a valid resource reference"
+    end
+
+    unless catalog.resource(res.to_s)
+      raise ArgumentError, "#{res} is not in the catalog"
+    end
```
Comment (reidmv) [inline lib/puppet/type/transition.rb]: We can't rely on this parameter value being a valid, in-the-catalog resource yet due to the possibility the referenced resource may not have been evaluated yet. I think that just means we shouldn't perform the `catalog.resource(res.to_s)` check.

Manifest:
```puppet
transition { 'stop myapp service':
  resource   => File['/tmp/test'],
  attributes => { ensure => absent },
  prior_to   => Notify['test']
}

file { '/tmp/test':
  ensure  => file,
  content => 'enabled=1',
  before  => Notify['test'],
}

notify { 'test': }
```

Error:
```
[reidmv@halcyon:~/src/puppetlabs-transition/] % puppet apply test.pp
Notice: Compiled catalog for halcyon.corp.puppetlabs.net in environment production in 0.05 seconds
Error: Parameter resource failed on Transition[stop myapp service]: File[/tmp/test] is not in the catalog at /Users/reidmv/src/puppetlabs-transition/test.pp:1
```

Comment (elyscape) [inline lib/puppet/type/transition.rb]: Hmm. In that case, maybe I should move the check to `pre_run_check`? We're guaranteed to have the whole catalog at that point.

File: lib/puppet/type/transition.rb (hunk 1 above)
Comment (elyscape) [inline lib/puppet/type/transition.rb]: Found the bug.

File: lib/puppet/type/transition.rb (hunk 2 above)
Comment (elyscape) [inline lib/puppet/type/transition.rb]: After doing some more research, I believe I can get all of this sort of validation to work so long as I move it into `pre_run_check`.

**Comment by elyscape** [comment]
@reidmv Since this pull request is reworking the validation, should I also remove [the type-wide `validate` method](https://github.com/puppetlabs/puppetlabs-transition/blob/0.1.0/lib/puppet/type/transition.rb#L115-L122), which only ensures that `resource`, `attributes`, and `prior_to` are provided, and just put `isrequired` on those attributes? Is there a reason you went that route instead of using `isrequired`?

**Comment by elyscape** [comment]
Okay, validation and munging moved to the `pre_run_check` method so that we can guarantee that resources will be in the catalog if they actually exist. Please weigh in on whether I should add `isrequired` to the `resource`, `attributes`, and `prior_to` attributes.

**Comment by reidmv** [comment]
@elyscape Thank you, this looks great! I'm :+1: to merge.

Regarding `isrequired`, I don't remember why I chose to do a global validate. I don't think it was functional, it would be fine to switch to just specifying `isrequired` on those properties/parameters. If you'd like to open a second PR on that I'll merge it.

**Comment by reidmv** [comment]
@hunner what are next steps for cutting a Forge release, given that transition in the puppetlabs namespace?

**Pull request merged**